### PR TITLE
[Feat] Migrate and add new NSA Forward Varlen operator(kernel)

### PR DESCRIPTION
**Pull request by wch0810**
## Description

Add Native Sparse Attention Forward with Variable-Length sequence support (NSA Forward VarLen) kernel for Ascend NPU, implemented in `example_tilelang_nsa_fwd_varlen.py` based on TileLang Developer mode (hybrid) — `pass_configs` four-True (combineCV + auto_cv_sync + auto_sync + memory_planning) with explicit L1/L0C/UB allocation, multi-buffer pipeline (num_stages=4) and persistent grid.

Single kernel (forward only, varlen). Developer mode hybrid: compiler auto-splits Cube/Vector scopes, auto-inserts cross-core flag sync (no `T.barrier_all`, no manual `T.Scope`, no manual `set_flag/wait_flag`). Layered test suite (21 cases) covering L0 threshold / L1 functional (incl. 4 model-typical configs) / L2 exception / Boundary special-value.

Tested on Ascend 910B3, fp16/bf16/fp32, N=2, C_SEQ=64, H=1, HQ=16, D=64, S=1, block_size=32. Precision verified against PyTorch CPU golden (single-pass softmax with S=1 specialization, mixed-tolerance dual-gate per `precision-standard.md` §4.1: float16 atol=2⁻¹⁴, rtol=2⁻⁹, max_abs_limit=1e-1, required_ratio=0.99, 21/21 PASS).

## Golden Config (N=2, C_SEQ=64, H=1, HQ=16, D=64, S=1, BS=32, fp16)

| Kernel | Task Duration (us) | Block Dim | Mix Block Dim | Max Diff |
|---|---|---|---|---|
| TileLang NSA Forward VarLen | 20.34 | 20 | 40 (20C+20V) | 4.883e-04 |

- 10 captures stable (19.94 ~ 20.66 us, ±1.8% spread, no outliers)
- Op Type: mix (CV fusion: Cube + Vector), Freq 1800/1800 MHz

## Precision (float16/bf16/fp32, precision-standard.md §4.1)

| Level | Cases | Result | matched_ratio | max_abs |
|---|---|---|---|---|
| L0 (blocking) | 1 | ✅ PASS | 1.0000 | 4.883e-04 |
| L1 (blocking) | 15 | ✅ PASS | 1.0000 | 0 ~ 1.953e-03 |
| L2 (non-blocking) | 2 | ✅ PASS | — (correctly rejected) | — |
| Boundary (non-blocking) | 3 | ✅ PASS | 0.9988 ~ 1.0000 | 0 ~ 9.766e-04 |

- L1 dtype coverage: fp16 (8 base + 4 typical), bf16, fp32, D=32
- L1 shape coverage: aligned (N=2,C=64), tail-1 (N=1,C=65), tail-mid (N=3,C=96), prime (N=3,C=97), edge (N=1,C=32)
- L1 typical configs: SEQ2K (N=2,C=2048), BATCH4 (N=4,C=1024), SEQ4K (N=1,C=4096), D128 (D=128)
- L2: float64 rejected, HQ=15 odd G rejected
- Boundary: zero (ratio=1.0), inf (ratio=0.9988), nan (ratio=0.9988)
- Dual-gate AND: matched_ratio ≥ 0.99 **AND** max_abs ≤ limit
- INF/NAN structural compare (§3.1)

## File Structure (2 files in this PR)

```
examples/deepseek_nsa/example_tilelang_nsa_fwd_varlen/
├── example_tilelang_nsa_fwd_varlen.py      # kernel + self-contained CI smoke test
└── test_example_tilelang_nsa_fwd_varlen.py  # golden + L0/L1/L2/Boundary precision tests
```

- `example_*.py`: `@tilelang.jit` kernel definition + `__main__` smoke test (single L0 case, self-contained, no test-file import, structural sanity check: shape + NaN/Inf + value range). Outputs "Test Passed!" for CI `bench_test.sh`.
- `test_*.py`: PyTorch golden (`naive_nsa_fwd_varlen`) + host helpers + `check_precision` (mixed-tolerance dual-gate) + L0/L1/L2/Boundary layered tests + `--level` dispatch (l0/l1/l2/boundary/all). Precision-only (no do_bench/msprof).

### CODE REVIEW

File: examples/deepseek_nsa/example_tilelang_nsa_fwd_varlen/example_tilelang_nsa_fwd_varlen.py [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+    if dtype_str != gemm_dtype_str:
+        q = q.to(gemm_dtype)
+        k = k.to(gemm_dtype)
+        v = v.to(gemm_dtype)
+
+    # GM workspace for Cube<->Vector data transfer (UB<->L1 not supported).
+    core_num = 20
+    num_stages = 4  # must match kernel
+
+    ws1 = torch.zeros(core_num, num_stages, G, BS, dtype=torch.float32, device="npu")
```
Comment (fengz72) [inline examples/deepseek_nsa/example_tilelang_nsa_fwd_varlen/example_tilelang_nsa_fwd_varlen.py]: 传参和workspace_idx冲突

File: examples/deepseek_nsa/example_tilelang_nsa_fwd_varlen/test_example_tilelang_nsa_fwd_varlen.py [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+# =============================================================================
+# Main: --level dispatch + exit code
+# =============================================================================
+def main():
+    parser = argparse.ArgumentParser(description="NSA Forward VarLen precision test suite (Ascend)")
+    parser.add_argument(
+        "--level",
+        default="l0",
+        choices=["l0", "l1", "l2", "boundary", "all"],
+        help="Test level to run (performance tests are in perf_example_tilelang_nsa_fwd_varlen.py)",
```
Comment (fengz72) [inline examples/deepseek_nsa/example_tilelang_nsa_fwd_varlen/test_example_tilelang_nsa_fwd_varlen.py]: perf_example_tilelang_nsa_fwd_varlen.py文件不存在

File: examples/deepseek_nsa/example_tilelang_nsa_fwd_varlen/example_tilelang_nsa_fwd_varlen.py [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+        V: T.Tensor([c_seq_len, head_kv, dim], gemm_dtype),  # type: ignore
+        # O_slc: final output = (acc_o / logsum) * g_slc (gate multiplication fused in kernel).
+        # Named O_slc for API compat with NSA paper; semantically equals golden's `o`
+        # (NOT golden's intermediate `o_slc` which excludes gate multiplication).
+        O_slc: T.Tensor([c_seq_len, heads, dim], dtype),  # type: ignore
+        BosPerToken: T.Tensor([c_seq_len], "int32"),  # type: ignore
+        IsSafe: T.Tensor([c_seq_len, head_kv], "int32"),  # type: ignore
+        BlockCounts: T.Tensor([c_seq_len, head_kv], "int32"),  # type: ignore
+        g_slc: T.Tensor([c_seq_len, heads], accum_dtype),  # type: ignore
+        Mask: T.Tensor([c_seq_len, head_kv, S, G, BS], accum_dtype),  # type: ignore
```
Comment (fengz72) [inline examples/deepseek_nsa/example_tilelang_nsa_fwd_varlen/example_tilelang_nsa_fwd_varlen.py]: 更常规的做法是 kernel 内用 BosPerToken/IsSafe/BlockCounts + token 位置算因果掩码

Comment (wch0810) [reply examples/deepseek_nsa/example_tilelang_nsa_fwd_varlen/example_tilelang_nsa_fwd_varlen.py]: 感谢建议！已按您的建议改为 kernel 内算因果掩码。

File: examples/deepseek_nsa/example_tilelang_nsa_fwd_varlen/test_example_tilelang_nsa_fwd_varlen.py (hunk 2 above)
Comment (wch0810) [reply examples/deepseek_nsa/example_tilelang_nsa_fwd_varlen/test_example_tilelang_nsa_fwd_varlen.py]: 感谢指出！已删除 test 文件对 `perf_example_tilelang_nsa_fwd_varlen.py` 的引用。

File: examples/deepseek_nsa/example_tilelang_nsa_fwd_varlen/example_tilelang_nsa_fwd_varlen.py (hunk 1 above)
Comment (wch0810) [reply examples/deepseek_nsa/example_tilelang_nsa_fwd_varlen/example_tilelang_nsa_fwd_varlen.py]: 感谢指出！已修复。移除了 host 端 `torch.zeros(...)` 创建 ws1/ws2/ws3 的代码，改用 `workspace_idx=[8, 9, 10]` 让框架自动分配。现在 kernel 调用时不再传入 workspace tensor，由框架在 `CythonKernelWrapper.forward` 中自动 `torch.empty(...)` 分配（`cython_wrapper.pyx:103-127`）。

**Review by fengz72 — APPROVED** [review]

**Review by MiaoHaifeng — APPROVED** [review]
/approve
